### PR TITLE
Add support for RFC 2579 DateAndTime

### PR DESCRIPTION
**Pull request by SuperQ**
Extract a UNIX timestamp from RFC 2579 DateAndTime textual conversions.
* Add support to the generator.
* Add support to the collector.
* Update snmp.yml with new DateAndTime metrics.

Fixes: https://github.com/prometheus/snmp_exporter/issues/321

Signed-off-by: Ben Kochie <superq@gmail.com>

### CODE REVIEW

File: generator/tree.go [hunk 1]
```diff
@@ -130,6 +137,8 @@ func metricType(t string) (string, bool) {
 		return "IpAddr", true
 	case "PhysAddress48", "DisplayString", "Float", "Double":
 		return t, true
+	case "DateAndTime":
```
[the remark this answers is not in the document]
Comment (SuperQ) [reply generator/tree.go]: This comes from the definition in the MIB/RFC.
```
DateAndTime ::= TEXTUAL-CONVENTION
```

[the remark this answers is not in the document]
Comment (RichiH) [reply generator/tree.go]: Ah, good.

[the remark this answers is not in the document]
Comment (brian-brazil) [reply generator/tree.go]: The textual convention is what we use for these, so this is fine.

File: collector.go [hunk 2]
```diff
@@ -234,6 +235,57 @@ func getPduValue(pdu *gosnmp.SnmpPDU) float64 {
...
+		return 0
+	}
+	pduLength := len(v)
+	// DateAndTime can be 8 or 11 bytes depending if the time zone is included.
+	switch pduLength {
+	case 8:
+		// No time zone included, assume UTC.
+		tz = time.UTC
+	case 11:
+		// Extract the timezone from hte last 3 bytes.
```
Comment (brian-brazil) [inline collector.go]: the

File: collector.go [hunk 3]
```diff
@@ -234,6 +235,57 @@ func getPduValue(pdu *gosnmp.SnmpPDU) float64 {
...
+			log.Debugf("error parsing location string: %q", locString)
+			return 0
+		}
+		tz = loc.Location()
+	default:
+		log.Debugf("invalid DateAndTime length %v", pduLength)
+		return 0
+	}
+	if err != nil {
+		log.Debugf("unable to parse DateAndTime %q", v)
```
Comment (brian-brazil) [inline collector.go]: Can you log the error?

File: collector.go [hunk 4]
```diff
@@ -234,6 +235,57 @@ func getPduValue(pdu *gosnmp.SnmpPDU) float64 {
...
+		tz  *time.Location
+		err error
+	)
+	// DateAndTime should be a slice of bytes.
+	switch pduType := pdu.Value.(type) {
+	case []byte:
+		v = pdu.Value.([]byte)
+	default:
+		log.Debugf("invalid DateAndTime type %v", pduType)
+		return 0
```
Comment (brian-brazil) [inline collector.go]: Should we really be returning 0 here?

File: generator/tree.go [hunk 5]
```diff
@@ -115,6 +115,13 @@ func prepareTree(nodes *Node) map[string]*Node {
 		}
 	})
 
+	// Convert RFC 2579 DateAndTime textual conversion to type.
+	walkNode(nodes, func(n *Node) {
```
Comment (brian-brazil) [inline generator/tree.go]: This and the previous two walks should really be combined

File: collector.go (hunk 4 above)
Comment (SuperQ) [reply collector.go]: I was debating how to do error handling here. The function that uses this, `pduToSamples`, doesn't really have any error handling.

File: generator/tree.go (hunk 5 above)
Comment (SuperQ) [reply generator/tree.go]: I saw this, and was wondering why we did walkNode over more than once. I can refactor.

File: collector.go (hunk 4 above)
Comment (brian-brazil) [reply collector.go]: We can either fail the scrape, or return nothing for this metric and log. Returning bad data is not a good option imho.

File: generator/README.md [hunk 6]
```diff
@@ -105,6 +105,7 @@ modules:
                              #   gauge:   An integer with type gauge.
                              #   counter: An integer with type counter.
                              #   OctetString: A bit string, rendered as 0xff34.
+                             #   DateAndTime: An RFC 2579 DateAndTime byte sequence.
```
Comment (brian-brazil) [inline generator/README.md]: I'd mention that UTC is presumed if the tz is missing.

**Comment by brian-brazil** [comment]
You need unittests for both the collector and generator changes.

**Comment by SuperQ** [comment]
@brian-brazil Yup, tests were why I left it WIP.

File: generator/FORMAT.md [hunk 7]
```diff
@@ -25,6 +26,10 @@ module_name:
      # as a label value on that gauge.
 
      # A metric that's part of a table, and thus has labels.
```
Comment (brian-brazil) [inline generator/FORMAT.md]: This isn't part of a table. You don't need to change this file.

Comment (SuperQ) [reply generator/FORMAT.md]: I was thinking it would be worth having an example in the FORMAT file. Don't care either way.

Comment (brian-brazil) [reply generator/FORMAT.md]: The format file is more the general structure, the various types are over in the readme.

Comment (SuperQ) [reply generator/FORMAT.md]: Got it, removed.

Comment (brian-brazil) [reply generator/FORMAT.md]: The types used to be in here, but they got moved when the type override was added. I'd prefer not to having to maintain two lists of types, as it's more likely they'll get out of sync.

**Pull request merged**